home/tests/test_signup.py

```diff
@@ -0,0 +1,71 @@
+import pytest
+from django.contrib.auth.models import User
+from home.models import Profile
+from pytest_django.asserts import assertTemplateUsed
+from home.forms import SignUpForm
+
+
+@pytest.mark.django_db
+class TestSignUp:
+
+    @pytest.fixture
+    def valid_signup_details(self):
+        return {'username': 'username', 'email': 'user@e.com', 'gender': 'U',
+                'password1': 'ido123123', 'password2': 'ido123123'}
+
+    def get_signup_page_status_code(self, client):
+        response = client.get('/signup')
+        assert response.status_code == 200
+
+    def test_valid_signup(self, valid_signup_details, client):
+        client.post('/signup/', data=valid_signup_details)
+        try:
+            user = User.objects.filter(username=valid_signup_details['username'])
+            assert user is not None
+            profile = Profile.objects.filter(user=user)
+            assert profile is not None
+        except ValueError:
+            assert False
+
+    @pytest.mark.parametrize("invalid_signup_details", [
+        # username cannot be empty
+        {'username': '', 'email': 'valid@email.com', 'gender': 'U',
+         'password1': 'pw123123', 'password2': 'pw123123'},
+        # email invalid
+        {'username': 'valid_username', 'email': 'a', 'gender': 'U',
+         'password1': 'pw123123', 'password2': 'pw123123'},
+        # password to short
+        {'username': 'valid_username', 'email': 'valid@email.com', 'gender': 'U',
+         'password1': '123', 'password2': '123'},
+        # passwords do not match
+        {'username': 'valid_username2', 'email': 'valid@email.com', 'gender': 'U',
+         'password1': '000000000', 'password2': '123456789'}
+    ])
+    def test_invalid_signup(self, invalid_signup_details, client):
+        client.post('/signup/', data=invalid_signup_details)
```

Comment:
don't we expect to get some kind of an error/retry page?

How is the error presented to the user?

Reply:
![image](https://user-images.githubusercontent.com/59822033/118613954-eb5c4600-b7c7-11eb-87c0-03d99d86859e.png)

Reply:
don't really know how check it, this is built in functionality of Django.
tried to check if there is some response when trying to press signup with invalid data, but there isn't.

Reply:
meanwhile checking that the page is not redirected (stay on signup page), and that the user is not created.

Reply:
The erros are included in the form object, see this example about how to render them manually to get an idea abut where they are:
https://docs.djangoproject.com/en/3.2/topics/forms/#rendering-form-error-messages

Reply:
Thanks, added tests.